lib/node/runner/block_operations.go

```diff
@@ -0,0 +1,236 @@
+	}
+}
+
+func (sb *SavingBlockOperations) CheckByBlock(st *storage.LevelDBBackend, blk block.Block) (err error) {
+	ops := make(chan string, 100)
+	results := make(chan error, 100)
+	defer close(results)
+
+	numWorker := int(len(blk.Transactions) / 2)
+	if numWorker > 100 {
+		numWorker = 100
+	}
+
+	for i := 1; i <= numWorker; i++ {
+		go sb.savingBlockOperationsWorker(i, st, blk, ops, results)
+	}
+	for _, txHash := range blk.Transactions {
+		ops <- txHash
+	}
+	close(ops)
+
+	var errs []error
+	for _ = range blk.Transactions {
+		err = <-results
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) > 0 {
+		err = errors.FailedToSaveBlockOperaton.Clone().SetData("errors", errs)
+		return
+	}
+
+	if blk.Height > common.GenesisBlockHeight { // ProposerTransaction
+		if err = sb.CheckTransactionByBlock(st, blk, blk.ProposerTransaction); err != nil {
+			return
+		}
+	}
+
+	return
+}
+
+func (sb *SavingBlockOperations) CheckTransactionByBlock(st *storage.LevelDBBackend, blk block.Block, hash string) (err error) {
+	var bt block.BlockTransaction
+	if bt, err = block.GetBlockTransaction(st, hash); err != nil {
+		sb.log.Error("failed to get BlockTransaction", "block", blk, "transaction", hash)
+		return
+	}
+
+	if bt.Transaction().IsEmpty() {
+		var tp block.TransactionPool
+		if tp, err = block.GetTransactionPool(st, hash); err != nil {
+			sb.log.Error("failed to get Transaction from TransactionPool", "transaction", hash)
+			return
+		}
+
+		bt.Message = tp.Message
+	}
+
+	for _, op := range bt.Transaction().B.Operations {
+		opHash := block.NewBlockOperationKey(op.MakeHashString(), hash)
+
+		var exists bool
+		if exists, err = block.ExistsBlockOperation(st, opHash); err != nil {
+			sb.log.Error(
+				"failed to check ExistsBlockOperation",
+				"block", blk,
+				"transaction", hash,
+				"operation", op,
+			)
+			return
+		}
+
+		if !exists {
+			bt.SaveBlockOperations(st, blk)
+			sb.log.Debug("saved missing BlockOperation", "block", blk, "transaction", hash, "operation", op)
+		}
+	}
+
+	return
+}
+
+func (sb *SavingBlockOperations) Start() {
+	go sb.continuousCheck()
+	go sb.StartSaving()
+
+	return
+}
+
+func (sb *SavingBlockOperations) StartSaving() {
+	sb.log.Debug("start saving")
+
+	for {
+		select {
+		case blk := <-sb.saveBlock:
+			if err := sb.save(blk); err != nil {
+				// NOTE if failed, the `continuousCheck()` will fill the missings.
+				sb.log.Error("failed to save BlockOperation", "block", blk, "error", err)
+			}
+		}
+	}
+}
+
+func (sb *SavingBlockOperations) Save(blk block.Block) {
+	go func() {
+		sb.saveBlock <- blk
+	}()
+}
+
+func (sb *SavingBlockOperations) save(blk block.Block) (err error) {
+	sb.log.Debug("start to save BlockOperation", "block", blk)
+	defer func() {
+		sb.log.Debug("end to save BlockOperation", "block", blk, "error", err)
+	}()
+
+	var st *storage.LevelDBBackend
+	if st, err = sb.st.OpenBatch(); err != nil {
+		return
+	}
+
+	if err = sb.CheckByBlock(st, blk); err != nil {
```

Comment:
Returned `err` from sb.CheckByBlock will be ignored in these two statements.
```
if err != nil {
   err = st.Discard()
} else {
   err = st.Commit()
}
```
Is this intended? But the returned error of `CheckByBlock` looks meaningful.